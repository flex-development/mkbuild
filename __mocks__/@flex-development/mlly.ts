/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const resolveAliases = vi.fn(actual.resolveAliases)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
