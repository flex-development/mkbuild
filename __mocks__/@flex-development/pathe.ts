/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
