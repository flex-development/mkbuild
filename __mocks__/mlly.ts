/**
 * @file Mocks - mlly
 * @module mocks/mlly
 * @see https://github.com/unjs/mlly
 */

/**
 * [`mlly`][1] module type.
 *
 * [1]: https://github.com/unjs/mlly
 */
type Actual = typeof import('mlly')

/**
 * `mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('mlly')

export const findDynamicImports = vi.fn(actual.findDynamicImports)
export const findExports = vi.fn(actual.findExports)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const resolvePath = vi.fn(actual.resolvePath)
