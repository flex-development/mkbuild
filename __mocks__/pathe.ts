/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
