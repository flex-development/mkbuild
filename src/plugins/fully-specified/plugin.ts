/**
 * @file Plugins - fully-specified
 * @module mkbuild/plugins/fully-specified
 */

import {
  BUILTIN_MODULES,
  EXT_JS_REGEX,
  EXT_TS_REGEX,
  RESOLVE_EXTENSIONS
} from '#src/config/constants'
import type { OutputMetadata } from '#src/types'
import extractStatements from '#src/utils/extract-statements'
import type {
  BuildOptions,
  BuildResult,
  OutputFile,
  Plugin,
  PluginBuild
} from 'esbuild'
import { resolvePath } from 'mlly'
import * as pathe from 'pathe'
import {
  readPackageJSON,
  resolvePackageJSON,
  type PackageJson
} from 'pkg-types'

/**
 * Plugin name.
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'fully-specified'

/**
 * Returns a specifier resolver plugin. There are three types of specifiers:
 *
 * > - *Relative specifiers* like `'./startup.js'` or `'../config.mjs'`. They
 * >   refer to a path relative to the location of the importing file. The file
 * >   extension is always necessary for these.
 * > - *Bare specifiers* like `'some-package'` or `'some-package/shuffle'`. They
 * >   can refer to the main entry point of a package by the package name, or a
 * >   specific feature module within a package prefixed by the package name as
 * >   per the examples respectively. Including the file extension is only
 * >   necessary for packages without an ["exports"][1] field.
 * > - *Absolute specifiers* like `'file:///opt/nodejs/config.js'`. They refer
 * >   directly and explicitly to a full path.
 *
 * The resolver adds file extensions to **bare** and **relative** specifiers in
 * output content.
 *
 * **Note**: [`--experimental-specifier-resolution=node`][2] can be used to
 * customize the ESM specifier resolution algorithm so that file extensions are
 * not required.
 *
 * [1]: https://nodejs.org/docs/latest-v16.x/api/packages.html#exports
 * [2]: https://nodejs.org/docs/latest-v16.x/api/esm.html#customizing-esm-specifier-resolution-algorithm
 *
 * @see https://nodejs.org/docs/latest-v16.x/api/esm.html#terminology
 *
 * @return {Plugin} Specifier resolver plugin
 */
const plugin = (): Plugin => {
  /**
   * Adds file extensions to relative specifiers in output file content.
   *
   * [1]: https://esbuild.github.io/plugins
   * [2]: https://esbuild.github.io/api/#build-api
   *
   * @param {PluginBuild} build - [esbuild plugin api][1]
   * @param {BuildOptions} build.initialOptions - [esbuild build api][2] options
   * @param {PluginBuild['onEnd']} build.onEnd - Build end callback
   * @param {PluginBuild['onStart']} build.onStart - Build start callback
   * @return {void} Nothing when complete
   */
  const setup = ({ initialOptions, onEnd, onStart }: PluginBuild): void => {
    const {
      absWorkingDir = process.cwd(),
      bundle,
      format = 'esm',
      metafile,
      outExtension: { '.js': ext = '.js' } = {},
      resolveExtensions: extensions = RESOLVE_EXTENSIONS
    } = initialOptions

    // bundle output shouldn't contain relative specifiers
    if (bundle) return

    // metafile required to get output metadata
    if (!metafile) {
      return void onStart(() => ({
        errors: [
          {
            detail: 'https://esbuild.github.io/api/#metafile',
            pluginName: PLUGIN_NAME,
            text: 'metafile required'
          }
        ]
      }))
    }

    return void onEnd(async (result: BuildResult): Promise<void> => {
      /**
       * Output file objects.
       *
       * @const {OutputFile[]} outputFiles
       */
      const outputFiles: OutputFile[] = []

      for (const output of result.outputFiles!) {
        /**
         * Output file extension.
         *
         * @const {string} output_ext
         */
        const output_ext: string = pathe.extname(output.path)

        // do nothing if output file isn't javascript or typescript
        if (!EXT_JS_REGEX.test(output_ext) && !EXT_TS_REGEX.test(output_ext)) {
          outputFiles.push(output)
          continue
        }

        /**
         * Relative path to output file.
         *
         * **Note**: Relative to {@link absWorkingDir}.
         *
         * @const {string} outfile
         */
        const outfile: string = output.path.replace(absWorkingDir + '/', '')

        /**
         * {@link output} metadata.
         *
         * @const {OutputMetadata} metadata
         */
        const metadata: OutputMetadata = result.metafile!.outputs[outfile]!

        // because this plugin doesn't handle bundles, the entry point can be
        // reset to the first (and only!) key in metadata.inputs
        if (!metadata.entryPoint) {
          metadata.entryPoint = Object.keys(metadata.inputs)[0]!
        }

        /**
         * Absolute path to source file.
         *
         * @const {string} source
         */
        const source: string = pathe.resolve(absWorkingDir, metadata.entryPoint)

        /**
         * {@link output.text} copy.
         *
         * @var {string} text
         */
        let text: string = output.text

        // replace specifiers
        for (const statement of extractStatements(output.text)) {
          /**
           * {@link statement.specifier} before file extension is added.
           *
           * @var {string | undefined} specifier
           */
          let specifier: string | undefined = statement.specifier

          // do nothing if missing module specifier
          if (!specifier) continue

          // do nothing if specifier references built-in module
          if (BUILTIN_MODULES.includes(specifier)) continue

          // do nothing if specifier is absolute
          if (/^(\/|(data|file|https?):)/.test(specifier)) continue

          // do nothing if specifier already includes file extension
          if (pathe.extname(specifier)) continue

          /**
           * {@link specifier} resolved.
           *
           * @see https://github.com/unjs/mlly#resolvepath
           *
           * @const {string} resolved
           */
          const resolved: string = await resolvePath(specifier, {
            conditions: [format === 'esm' ? 'import' : 'require'],
            extensions,
            url: specifier.startsWith('.') ? source : pathe.dirname(source)
          })

          // add extension to bare or relative specifier
          // bare specifier => resolved contains '/node_modules/'
          // relative specifier => resolved does not contain '/node_modules/'
          if (/\/node_modules\//.test(resolved)) {
            /**
             * Path to `package.json` of package {@link resolved} references.
             *
             * @const {string} pkgfile
             */
            const pkgfile: string = await resolvePackageJSON(resolved, {
              startingFrom: pathe.dirname(resolved)
            })

            /**
             * `package.json` of package {@link resolved} references.
             *
             * @const {PackageJson} pkg
             */
            const pkg: PackageJson = await readPackageJSON(pkgfile)

            /**
             * {@link pkg} entry point resolved.
             *
             * @const {string} main
             */
            const main: string = pathe.resolve(
              absWorkingDir,
              'node_modules',
              pkg.name!,
              pkg.main!
            )

            // update specifier if pkg does not have exports field
            // and resolved is not pkg entry point
            if (pkg.exports === undefined && resolved !== main) {
              specifier = resolved.replace(/.+(node_modules\/)/, '')
            }
          } else {
            // get basename of specifier and resolved
            const { name } = pathe.parse(specifier)
            const { name: resolved_name } = pathe.parse(resolved)

            // specifier resolved to a directory
            if (name !== resolved_name) specifier += '/' + resolved_name

            // add output file extension to specifier
            specifier += ext
          }

          // replace specifier
          text = text.replace(
            statement.code,
            statement.code.replace(statement.specifier!, specifier)
          )
        }

        outputFiles.push({
          ...output,
          contents: new Uint8Array(Buffer.from(text)),
          text
        })
      }

      return void (result.outputFiles = outputFiles)
    })
  }

  return { name: PLUGIN_NAME, setup }
}

export default plugin
