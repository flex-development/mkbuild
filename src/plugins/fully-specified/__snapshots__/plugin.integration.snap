// Vitest Snapshot v1

exports[`integration:plugins/fully-specified > esbuild > should fill specifiers in cjs output files 1`] = `
"require(\\"esbuild\\");
require(\\"node:util\\");
require(\\"pathe\\");
require(\\"tsconfig-paths\\");
require(\\"tsconfig-paths/lib/tsconfig-loader.js\\");
require(\\"../src/config/constants.ts\\");
require(\\"../src/interfaces/index.cjs\\");
"
`;

exports[`integration:plugins/fully-specified > esbuild > should fill specifiers in copied output files 1`] = `
"/**
 * @file Fixtures - my-atoi.d.mts
 * @module fixtures/my-atoi.d.mts
 */

/**
 * Converts a string, \`s\`, to a 32-bit signed integer (similar to C/C++'s \`atoi\`
 * function).
 *
 * @example myAtoi('') // 0
 * @example myAtoi('42') // 42
 * @example myAtoi('   -42') // -42
 * @example myAtoi('4193 with words') // 4193
 * @example myAtoi('words and 987') // 0
 * @example myAtoi('+-12') // 0
 *
 * @param {string} s - String to convert
 * @return {number} 32-bit signed integer
 */
declare function myAtoi(s: string): number

export default myAtoi

export { default as MAX_VALUE } from './max-value.mjs'
export { default as MIN_VALUE } from './min-value.mjs'
"
`;

exports[`integration:plugins/fully-specified > esbuild > should fill specifiers in esm output files 1`] = `
"import \\"esbuild\\";
import \\"node:util\\";
import \\"pathe\\";
import \\"tsconfig-paths\\";
import \\"tsconfig-paths/lib/tsconfig-loader.js\\";
import \\"../src/config/constants.ts\\";
import \\"../src/interfaces/index.mjs\\";
"
`;

exports[`integration:plugins/fully-specified > esbuild > should skip files that are not javascript or typescript 1`] = `
"{
  currency: 'USD',
  prices: [
    { price: '2486.69', time: '2017-07-03T00:00:00Z' },
    { price: '2457.05', time: '2017-07-02T00:00:00Z' },
    { price: '2486.09', time: '2017-07-01T00:00:00Z' },
    { price: '2535.85', time: '2017-06-30T00:00:00Z' },
    { price: '2546.06', time: '2017-06-29T00:00:00Z' }
  ]
}
"
`;
